src/hotspot/share/interpreter/zero/bytecodeInterpreter.cpp

```diff
@@ -2415,7 +2452,7 @@ void BytecodeInterpreter::run(interpreterState istate) {
             CHECK_NULL(STACK_OBJECT(-(cache->parameter_size())));
             if (cache->is_vfinal()) {
               callee = cache->f2_as_vfinal_method();
-              if (REWRITE_BYTECODES) {
+              if (REWRITE_BYTECODES && !UseSharedSpaces && !Arguments::is_dumping_archive()) {
```

Comment:
As we discussed with @tstuefe in another thread, the use of `UseSharedSpaces` needs the explicit `#include` of `globals.hpp`.

Reply:
When running with dynamic dump (`-XX:ArchiveClassesAtExit=foo.jsa`), `UseSharedSpaces` is true, so it's possible for a method to be rewritten here, and later dumped into the CDS archive). 

I think we should remove `!UseSharedSpaces`. I'll try to write a test case for it. If I understand correctly, even if `cache->is_vfinal()` is true at dump time, it's not guarantee to be true at run time (we might load a different version of the class that contains the target method).

Reply:
I thought if we run with dynamic dump, then `Arguments::is_dumping_archive()` is `false`, and we don't rewrite.

Anyway, if you remove `!UseSharedSpaces` here, like you did in PR update, then this happens during Linux x86_64 Zero build:

```
$ CONF=linux-x86_64-zero-fastdebug make images
...
Creating support/classlist.jar
/bin/bash: line 14: 2355592 Aborted                 (core dumped) /home/shade/trunks/jdk/build/linux-x86_64-zero-fastdebug/support/interim-image/bin/java ... x86_64-zero-fastdebug/support/link_opt/default_jli_trace.txt
ERROR: Failed to generate link optimization data. This is likely a problem with the newly built JVM/JDK.
# To suppress the following error report, specify this argument
# after -XX: or in .hotspotrc:  SuppressErrorAt=/os_linux_zero.cpp:90
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/shade/trunks/jdk/src/hotspot/os_cpu/linux_zero/os_linux_zero.cpp:90), pid=2355592, tid=2355593
#  Error: ShouldNotCall()

```

That is, current PR breaks Zero.

My code was mimicking what `TemplateTable::invokevirtual` in PPC64 did:

```
  if (RewriteBytecodes && !UseSharedSpaces && !DumpSharedSpaces) {
    patch_bytecode(Bytecodes::_fast_invokevfinal, Rnew_bc, R12_scratch2);
  }
```

Reply:
You're right. I added the `!UseSharedSpaces` back.  I misread the code :-(